storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java

```diff
@@ -0,0 +1,553 @@
+     */
+    protected int smallestUpperBoundSlotFor(ByteBuffer idx, long target, IndexSearchType searchEntity) {
+        return indexSlotRangeFor(idx, target, searchEntity).smallestUpperBound;
+    }
+
+    /**
+     * Round a number to the greatest exact multiple of the given factor less than the given number.
+     * E.g. roundDownToExactMultiple(67, 8) == 64
+     */
+    private static int roundDownToExactMultiple(int number, int factor) {
+        return factor * (number / factor);
+    }
+
+    private static MappedByteBuffer createMappedBuffer(RandomAccessFile raf, boolean newlyCreated, long length,
+                                                       boolean writable, int entrySize) throws IOException {
+        MappedByteBuffer idx;
+        if (writable)
+            idx = raf.getChannel().map(FileChannel.MapMode.READ_WRITE, 0, length);
+        else
+            idx = raf.getChannel().map(FileChannel.MapMode.READ_ONLY, 0, length);
+
+        /* set the position in the index for the next entry */
+        if (newlyCreated)
+            idx.position(0);
+        else
+            // if this is a pre-existing index, assume it is valid and set position to last entry
+            idx.position(roundDownToExactMultiple(idx.limit(), entrySize));
+
+        return idx;
+    }
+
+    /**
+     * Lookup lower and upper bounds for the given target.
+     */
+    private BinarySearchResult indexSlotRangeFor(ByteBuffer idx, long target, IndexSearchType searchEntity) {
+        // check if the index is empty
+        if (entries == 0)
+            return new BinarySearchResult(-1, -1);
+
+        int firstHotEntry = Math.max(0, entries - 1 - warmEntries());
+        // check if the target offset is in the warm section of the index
+        if (compareIndexEntry(parseEntry(idx, firstHotEntry), target, searchEntity) < 0) {
+            return binarySearch(idx, target, searchEntity, firstHotEntry, entries - 1);
+        }
+
+        // check if the target offset is smaller than the least offset
+        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0)
+            return new BinarySearchResult(-1, 0);
+
+        return binarySearch(idx, target, searchEntity, 0, firstHotEntry);
+    }
+
+    private BinarySearchResult binarySearch(ByteBuffer idx, long target, IndexSearchType searchEntity, int begin, int end) {
+        // binary search for the entry
+        int lo = begin;
+        int hi = end;
+        while (lo < hi) {
+            int mid = (lo + hi + 1) >>> 1;
+            IndexEntry found = parseEntry(idx, mid);
+            int compareResult = compareIndexEntry(found, target, searchEntity);
+            if (compareResult > 0)
+                hi = mid - 1;
+            else if (compareResult < 0)
+                lo = mid;
+            else
+                return new BinarySearchResult(mid, mid);
+        }
+        int blah;
```

Comment:
Instead of defining blah, could we just reuse hi?

Reply:
My bad, I meant to rename this, but missed it. Your suggestion sounds good.